src/Microsoft.ML.Core/Data/IEstimator.cs

```diff
@@ -16,13 +17,17 @@ namespace Microsoft.ML.Core.Data
     /// This is more relaxed than the proper <see cref="ISchema"/>, since it's only a subset of the columns,
     /// and also since it doesn't specify exact <see cref="ColumnType"/>'s for vectors and keys.
     /// </summary>
-    public sealed class SchemaShape
+    public sealed class SchemaShape : IReadOnlyList<SchemaShape.Column>
     {
         public readonly Column[] Columns;
```

Comment:
>Column [](start = 24, length = 6)

Can you make it ImmutableList?
What will save you from IEnumarable casting in line 205

Reply:
What type is sealed? ImmuatableList? Why it's bad?

Reply:
I am not sure if I understand your comment correctly. Do you mean I need to use `public class SchemaShape : ImmutableList<SchemaShape.Column>,` which can't not be compiled because classes can't be derived from sealed classes?

---
In reply to: [238881290](https://github.com/dotnet/machinelearning/pull/1822#discussion_r238881290) [](ancestors = 238881290)